### PR TITLE
refactor(kad): use message patterns from `asynchronous-codec`

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Notes & open questions

This is a PoC for the message patterns prototyped in https://github.com/mxinden/asynchronous-codec/pull/5.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@mxinden Can you take a look at this and let me know what you think? These "message patterns" would also be useful in gossipsub and a lot of other protocols where we currently write these state machines manually.

File: protocols/kad/src/indexed_stream.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    S: Stream,
+{
+    type Item = (I, S::Item);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Safety: We never expose an unpinned reference.
+        let index = self.index.clone();
+
+        match futures::ready!(unsafe {
+            Pin::new_unchecked(&mut self.get_unchecked_mut().stream).poll_next(cx)
```
Comment (mxinden) [inline protocols/kad/src/indexed_stream.rs:37]: Why not use pin project? Why risk being wrong with the `Pin` behavior here?

File: protocols/kad/src/indexed_stream.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    pub fn index(&self) -> &I {
+        &self.index
+    }
+
+    pub fn stream_pin_mut(&mut self) -> Pin<&mut S> {
+        // Safety: We never expose an unpinned reference.
+        unsafe { Pin::new_unchecked(&mut self.stream) }
+    }
```
Comment (mxinden) [inline protocols/kad/src/indexed_stream.rs:22]: Same here, why not use pin project?

File: protocols/kad/src/protocol.rs [hunk 3]
```diff
@@ -184,95 +182,70 @@ impl<C> InboundUpgrade<C> for KademliaProtocolConfig
...
-                .and_then::<_, fn(_) -> _>(|bytes| {
-                    let mut reader = BytesReader::from_bytes(&bytes);
-                    let request = match proto::Message::from_reader(&mut reader, &bytes) {
-                        Ok(r) => r,
-                        Err(err) => return future::ready(Err(err.into())),
-                    };
-                    future::ready(proto_to_req_msg(request))
-                }),
-        )
+        future::ok(incoming)
```
Comment (mxinden) [inline protocols/kad/src/protocol.rs:190]: Why not use a `ReadyUpgrade` here?

**Review by mxinden — COMMENTED** [review]
Direction looks good to me. I think we can easily get around the `unsafe` with `pin-project`.

File: protocols/kad/src/indexed_stream.rs (hunk 1 above)
Comment (thomaseizinger) [reply protocols/kad/src/indexed_stream.rs:37]: Just lazy and typing it was quicker in the PoC. Wasn't sure if I would actually need the type permanently.

**Comment by thomaseizinger** [comment]
> Direction looks good to me.

Cool! I'll then go ahead and polish the `asynchronous-codec` PR first. I want to see how much easier they would be to implement with `async/await`.

File: protocols/kad/src/protocol.rs (hunk 3 above)
Comment (thomaseizinger) [reply protocols/kad/src/protocol.rs:190]: `ReadyUpgrade` only supports a single protocol name.

**Pull request closed without merging**